### PR TITLE
turn: block forwarding to loopback/any

### PR DESCRIPTION
**Pull request by z-dule**

### CODE REVIEW

File: modules/turn/turn.c [hunk 1]
```diff
@@ -177,13 +178,17 @@ static bool indication_handler(struct restund_msgctx *ctx, int proto,
...
-	perm = perm_find(al->perms, &peer->v.xor_peer_addr);
+	psa = &peer->v.xor_peer_addr;
+	perm = perm_find(al->perms, psa);
 	if (!perm) {
 		++al->dropc_tx;
 		return true;
 	}
 
-	err = udp_send(al->rel_us, &peer->v.xor_peer_addr, &data->v.data);
+	if (sa_is_loopback(psa) || sa_is_any(psa))
```
Comment (julialongtin) [inline modules/turn/turn.c]: also sa_is_linklocal?

Comment (arianvp) [reply modules/turn/turn.c]: I think just `127.0.0.1` and `0.0.0.0` is fine for now.   But it doesn't _hurt_

**Review by julialongtin — APPROVED** [review]

**Review by franziskuskiefer — COMMENTED** [review]
lgtm if these are all the places this has to happen.
Do we have a way to test this?

File: modules/turn/turn.c [hunk 2]
```diff
@@ -177,13 +178,17 @@ static bool indication_handler(struct restund_msgctx *ctx, int proto,
...
-	perm = perm_find(al->perms, &peer->v.xor_peer_addr);
+	psa = &peer->v.xor_peer_addr;
+	perm = perm_find(al->perms, psa);
 	if (!perm) {
 		++al->dropc_tx;
 		return true;
 	}
 
-	err = udp_send(al->rel_us, &peer->v.xor_peer_addr, &data->v.data);
+	if (sa_is_loopback(psa) || sa_is_any(psa) || sa_is_linklocal(psa))
```
Comment (arianvp) [inline modules/turn/turn.c]: Hmm `sa_is_loopback` checks if the address is equal to `127.0.0.1` but  shouldn't we check whether the address is in the  `127.0.0.0/8` range? This seems to be a bug in the upstream `re` library we're using.

There are other things on loopback that we do not want to be reachable. E.g. `127.0.0.53` hosts the `systemd-resolved` DNS resolver  on Ubuntu machines and relaying to it could trigger a DoS


re: https://github.com/creytiv/re/blob/ea0509d6fc9a8beb366dde731885ff7d9b4d2d8b/src/sa/sa.c#L555

coturn: https://github.com/coturn/coturn/blob/060bf187879fd1a6386012f4c5a7494824ebe5c8/src/client/ns_turn_ioaddr.c#L478-L497

**Review by arianvp — CHANGES_REQUESTED** [review]

File: modules/turn/turn.c (hunk 2 above)
Comment (z-dule) [reply modules/turn/turn.c]: Good points. Added.

**Review by arianvp — APPROVED** [review]

**Comment by arianvp** [comment]
Using coturn's testing tools I'm checking if the relaying is being blocked. On the current build of  restund indeed still relaying the addresses.:
```
$ nix run nixpkgs.coturn
$ turnutils_uclient  -X 192.168.0.136 -u test -w secret -e 127.0.0.2 & turnutils_uclient  -X 192.168.0.136 -u test -w secret -e 255.255.255.255 &  turnutils_uclient  -X 192.168.0.136 -u test -w secret -e 169.254.0.1

```


`turn` command on `status` interface shows 127.0.0.2, 255.255.255.255 and  169.254.0.1  being relayed
```
TURN relay=192.168.0.136 relay6=? (err 5/0)
- 34295 UDP/192.168.0.136:34159/192.168.0.136:3478 - 192.168.0.136:38566 "test" 594s (drop 0/0)
    permissions: (169.254.0.1 295s relay 5/0)
    channels:    (0x59c7 169.254.0.1:3481 595s)
- 37255 UDP/192.168.0.136:37119/192.168.0.136:3478 - 192.168.0.136:35496 "test" 655s (drop 0/0)
    permissions:
    channels:   
- 37268 UDP/192.168.0.136:37132/192.168.0.136:3478 - 192.168.0.136:34725 "test" 594s (drop 0/0)
    permissions: (127.0.0.2 295s relay 5/0)
    channels:    (0x6db7 127.0.0.2:3481 595s) (0x6fce 127.0.0.2:3480 595s)
- 37472 UDP/192.168.0.136:37336/192.168.0.136:3478 - 192.168.0.136:34724 "test" 771s (drop 0/0)
    permissions:
    channels:   
- 42726 UDP/192.168.0.136:42590/192.168.0.136:3478 - 192.168.0.136:35114 "test" 156s (drop 0/0)
    permissions:
    channels:   
- 43796 UDP/192.168.0.136:43660/192.168.0.136:3478 - 192.168.0.136:52394 "test" 594s (drop 0/0)
    permissions: (255.255.255.255 295s relay 0/0)
    channels:    (0x701e 255.255.255.255:3481 595s)
- 47256 UDP/192.168.0.136:47120/192.168.0.136:3478 - 192.168.0.136:52854 "test" 594s (drop 0/0)
    permissions: (127.0.0.2 295s relay 5/0)
    channels:    (0x6719 127.0.0.2:3481 595s)
- 47291 UDP/192.168.0.136:47155/192.168.0.136:3478 - 192.168.0.136:43029 "test" 594s (drop 0/0)
    permissions: (169.254.0.1 295s relay 5/0)
    channels:    (0x45b8 169.254.0.1:3481 595s) (0x54fb 169.254.0.1:3480 595s)
- 52136 UDP/192.168.0.136:52000/192.168.0.136:3478 - 192.168.0.136:47978 "test" 631s (drop 0/0)
    permissions:
    channels:   
- 52855 UDP/192.168.0.136:52719/192.168.0.136:3478 - 192.168.0.136:43028 "test" 771s (drop 0/0)
    permissions:
    channels:   
- 54504 UDP/192.168.0.136:54368/192.168.0.136:3478 - 192.168.0.136:37080 "test" 771s (drop 0/0)
    permissions:
    channels:   
- 55756 UDP/192.168.0.136:55620/192.168.0.136:3478 - 192.168.0.136:37081 "test" 594s (drop 0/0)
    permissions: (255.255.255.255 295s relay 0/0)
    channels:    (0x77f2 255.255.255.255:3480 595s) (0x5455 255.255.255.255:3481 595s)
- 57837 UDP/192.168.0.136:57701/192.168.0.136:3478 - 192.168.0.136:46828 "test" 179s (drop 0/0)
    permissions:
    channels:  
```

I will now make a new build and see if the issue is mitigated

File: modules/turn/turn.c [hunk 3]
```diff
@@ -177,13 +194,17 @@ static bool indication_handler(struct restund_msgctx *ctx, int proto,
...
+	psa = &peer->v.xor_peer_addr;
+	perm = perm_find(al->perms, psa);
 	if (!perm) {
 		++al->dropc_tx;
 		return true;
 	}
 
-	err = udp_send(al->rel_us, &peer->v.xor_peer_addr, &data->v.data);
+	if (is_blocked(psa))
+		err = EPERM;
```
Comment (arianvp) [inline modules/turn/turn.c:205]: Shouldn't we return here? Otherwise

**Review by arianvp — CHANGES_REQUESTED** [review]
@z-dule  I tried to see if the issue at hand is fixed with this build; but I'm still able to succesfully open a channel-bind on `127.0.0.1`. 

image

Maybe I'm missing something here. but it doesn't seem fixed.

**Comment by arianvp** [comment]
Aaaah looking at the code where we add the check we're already past the channel-bind and at the "send data packets" part and we simply drop the packets on the floor. Is that correct? 

But could we also add the check  to https://github.com/wireapp/restund/blob/edd4abd92f2a8b08a4ef4b704b84dcb13fb28231/modules/turn/chan.c#L232 so that the channel-bind actually fails instead of accepting and silently dropping packets?

In short; also add the check to `request_handler`; not just `raw_handler` and `indication_handler`

would that make sense?


NOTE: This is me not knowing a lot about TURN so please tell me when I'm saying stupid things =)

**Review by franziskuskiefer — APPROVED** [review]
